File: src/main/java/com/github/spotbugs/snom/SpotBugsExtension.java
/**
 * Copyright 2019 SpotBugs team
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.spotbugs.snom;

import edu.umd.cs.findbugs.annotations.NonNull;
import edu.umd.cs.findbugs.annotations.Nullable;
import java.io.File;
import java.nio.file.Paths;
import java.util.Collection;
import javax.inject.Inject;
import org.gradle.api.Project;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.ListProperty;
import org.gradle.api.provider.Property;

public class SpotBugsExtension {
  @NonNull final Property<Boolean> ignoreFailures;
  @NonNull final Property<Boolean> showProgress;
  @NonNull final Property<Confidence> reportLevel;
  @NonNull final Property<Effort> effort;
  @NonNull final ListProperty<String> visitors;
  @NonNull final ListProperty<String> omitVisitors;
  @NonNull final Property<File> reportsDir;
  @NonNull final Property<File> includeFilter;
  @NonNull final Property<File> excludeFilter;
  @NonNull final ListProperty<String> onlyAnalyze;
  @NonNull final Property<String> projectName;
  @NonNull final Property<String> release;

  @Inject
  public SpotBugsExtension(Project project, ObjectFactory objects) {
    ignoreFailures = objects.property(Boolean.class);
    showProgress = objects.property(Boolean.class);
    reportLevel = objects.property(Confidence.class);
    effort = objects.property(Effort.class);
    visitors = objects.listProperty(String.class);
    omitVisitors = objects.listProperty(String.class);
    reportsDir = objects.property(File.class);
    // the default reportsDir is "$buildDir/reports/spotbugs"
    reportsDir.set(
        project.getBuildDir().toPath().resolve(Paths.get("reports", "spotbugs")).toFile());
    includeFilter = objects.property(File.class);
    excludeFilter = objects.property(File.class);
    onlyAnalyze = objects.listProperty(String.class);
    projectName = objects.property(String.class);
    projectName.convention(project.getName());
    release = objects.property(String.class);
  }

  public Property<File> getReportsDir() {
    return reportsDir;
  }

  public void setReportsDir(Property<File> file) {
    reportsDir.set(file);
  }

  public void setReportsDir(File file) {
    reportsDir.set(file);
  }

  public Property<File> getIncludeFilter() {
    return includeFilter;
  }

  public void setIncludeFilter(Property<File> file) {
    includeFilter.set(file);
  }

  public void setIncludeFilter(@Nullable File file) {
    includeFilter.set(file);
  }

  public Property<File> getExcludeFilter() {
    return excludeFilter;
  }

  public void setExcludeFilter(Property<File> file) {
    excludeFilter.set(file);
  }

  public void setExcludeFilter(@Nullable File file) {
    excludeFilter.set(file);
  }

  public void setVisitors(@Nullable Collection<String> collection) {
    visitors.set(collection);
  }

  public void setOmitVisitors(@Nullable Collection<String> collection) {
    omitVisitors.set(collection);
  }

  public void setOnlyAnalyze(@Nullable Collection<String> collection) {
    onlyAnalyze.set(collection);
  }
}

File: src/main/kotlin/com/github/spotbugs/snom/SpotBugsExtension.kt
/*
 * Copyright 2021 SpotBugs team
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.spotbugs.snom

import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.Property

/**
 * [SpotBugsExtension] is an extension used to set up the SpotBugs Gradle plugin.
 * All properties in this extension act as default properties for all instances of [SpotBugsTask] are optional.
 *
 * ### Usage
 * Once you've applied the SpotBugs Gradle plugin to your project, configure it as shown below:
 *
 * ```kotlin
 * // Required: Gradle 8.2 or higher
 * spotbugs {
 *     ignoreFailures = false
 *     showStackTraces = true
 *     showProgress = true
 *     effort = com.github.spotbugs.snom.Effort.DEFAULT
 *     reportLevel = com.github.spotbugs.snom.Confidence.DEFAULT
 *     visitors = listOf("FindSqlInjection", "SwitchFallthrough")
 *     omitVisitors = listOf("FindNonShortCircuit")
 *     reportsDir = file("$buildDir/spotbugs")
 *     includeFilter = file("include.xml")
 *     excludeFilter = file("exclude.xml")
 *     baselineFile = file("baseline.xml")
 *     onlyAnalyze = listOf("com.foobar.MyClass", "com.foobar.mypkg.*")
 *     maxHeapSize = "1g"
 *     extraArgs = listOf("-nested:false")
 *     jvmArgs = listOf("-Duser.language=ja")
 *     runOnCheck = true
 * }
 * ```
 *
 * See also [SpotBugs Manual about configuration](https://spotbugs.readthedocs.io/en/stable/running.html).
 */
interface SpotBugsExtension {
    val ignoreFailures: Property<Boolean>
    val showStackTraces: Property<Boolean>

    /**
     * Property to enable progress reporting during the analysis. Default value is `false`.
     */
    val showProgress: Property<Boolean>

    /**
     * Property to specify the level to report bugs. Default value is [Confidence.DEFAULT].
     */
    val reportLevel: Property<Confidence>

    /**
     * Property to adjust SpotBugs detectors. Default value is [Effort.DEFAULT].
     */
    val effort: Property<Effort>

    /**
     * Property to enable visitors (detectors) for analysis. Default is empty that means all visitors run analysis.
     */
    val visitors: ListProperty<String>

    /**
     * Property to disable visitors (detectors) for analysis. Default is empty that means SpotBugs omits no visitor.
     */
    val omitVisitors: ListProperty<String>

    /**
     * Property to set the directory to generate report files. Default is `"$buildDir/reports/spotbugs"`.
     *
     * Note that each [SpotBugsTask] creates own subdirectory in this directory.
     */
    val reportsDir: DirectoryProperty

    /**
     * Property to set the filter file to limit which bug should be reported.
     *
     * Note that this property will NOT limit which bug should be detected. To limit the target classes to analyze,
     * use [onlyAnalyze] instead.
     * To limit the visitors (detectors) to run, use [visitors] and [omitVisitors] instead.
     *
     * See also [SpotBugs Manual about Filter file](https://spotbugs.readthedocs.io/en/stable/filter.html).
     */
    val includeFilter: RegularFileProperty

    /**
     * Property to set the filter file to limit which bug should be reported.
     *
     * Note that this property will NOT limit which bug should be detected. To limit the target classes to analyze,
     * use [onlyAnalyze] instead.
     * To limit the visitors (detectors) to run, use [visitors] and [omitVisitors] instead.
     *
     * See also [SpotBugs Manual about Filter file](https://spotbugs.readthedocs.io/en/stable/filter.html).
     */
    val excludeFilter: RegularFileProperty

    /**
     * Property to set the baseline file. This file is a Spotbugs result file, and all bugs reported in this file
     * will not be reported in the final output.
     */
    val baselineFile: RegularFileProperty

    /**
     * Property to specify the target classes for analysis. Default value is empty that means all classes are analyzed.
     */
    val onlyAnalyze: ListProperty<String>

    /**
     * Property to specify the name of project. Some reporting formats use this property.
     * Default value is the name of your Gradle project.
     */
    val projectName: Property<String>

    /**
     * Property to specify the release identifier of project. Some reporting formats use this property.
     * Default value is the version of your Gradle project.
     */
    val release: Property<String>

    /**
     * Property to specify the extra arguments for SpotBugs.
     * Default value is empty so SpotBugs will get no extra argument.
     */
    val extraArgs: ListProperty<String>

    /**
     * Property to specify the extra arguments for JVM process. Default value is empty so JVM process will get
     * no extra argument.
     */
    val jvmArgs: ListProperty<String>

    /**
     * Property to specify the max heap size (`-Xmx` option) of JVM process.
     * Default value is empty so the default configuration made by Gradle will be used.
     */
    val maxHeapSize: Property<String>

    val toolVersion: Property<String>

    val useAuxclasspathFile: Property<Boolean>

    val useJavaToolchains: Property<Boolean>

    /**
     * Property to specify if the SpotBugs tasks should automatically be marked as dependencies of the
     * check task. Defaults to true.
     */
    val runOnCheck: Property<Boolean>
}
